lib/portlayer/storage/container/import.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package container
+
+import (
+	"errors"
+	"io"
+	"net/url"
+	"os"
+
+	"github.com/vmware/vic/lib/archive"
+	"github.com/vmware/vic/lib/guest"
+	"github.com/vmware/vic/lib/portlayer/storage"
+	"github.com/vmware/vic/lib/portlayer/storage/vsphere"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/disk"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+func (c *ContainerStore) Import(op trace.Operation, id string, spec *archive.FilterSpec, tarstream io.ReadCloser) error {
+	l, err := c.NewDataSink(op, id)
+	if err != nil {
+		return err
+	}
+
+	return l.Import(op, spec, tarstream)
+}
+
+// NewDataSink creates and returns an DataSink associated with container storage
+func (c *ContainerStore) NewDataSink(op trace.Operation, id string) (storage.DataSink, error) {
+	uri, err := c.URL(op, id)
+	if err != nil {
+		return nil, err
+	}
+
+	offlineAttempt := 0
+offline:
+	offlineAttempt++
+
+	sink, err := c.newDataSink(op, uri)
+	if err == nil {
+		return sink, err
+	}
+
+	// check for vmdk locked error here
+	if !disk.IsLockedError(err) {
+		op.Warnf("Unable to mount %s and do not know how to recover from error")
+		// continue anyway because maybe there's an online option
+	}
+
+	// online - Owners() should filter out the appliance VM
+	// #nosec: Errors unhandled.
+	owners, _ := c.Owners(op, uri, disk.LockedVMDKFilter)
+	if len(owners) == 0 {
+		op.Infof("No online owners were found for %s", id)
+		return nil, errors.New("unable to create offline data sink and no online owners found")
+	}
+
+	for _, o := range owners {
+		// sanity check to see if we are the owner - this should catch transitions
+		// from container running to diff or commit for example between the offline attempt and here
+		uuid, err := o.UUID(op)
+		if err == nil {
+			// check if the vm is appliance VM if we can successfully get its UUID
+			// #nosec: Errors unhandled.
+			self, _ := guest.IsSelf(op, uuid)
+			if self && offlineAttempt < 2 {
+				op.Infof("Appliance is owner of online vmdk - retrying offline source path")
+				goto offline
```

Comment:
Curious: do we want to retry without any sort of time delay/backoff?

Reply:
Yes - this code has determined that the disk in question is already connected to the appliance and that's why we couldn't attach it. That means we should be able to fall back to the offline path immediately.
There's no repeated retry so no delay/backoff logic.